File: tg-bulk-invite/src/pages/api/telegram/getParticipants.ts
import type { NextApiRequest, NextApiResponse } from 'next';
import { Api } from "telegram";
import { activeClients } from './connect';

type ResponseData = {
  success: boolean;
  message: string;
  participants?: any[];
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<ResponseData>
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ success: false, message: 'Method not allowed' });
  }

  const { sourceGroups, targetGroup, sessionId } = req.body;

  try {
    const clientData = activeClients.get(sessionId);
    if (!clientData) {
      return res.status(400).json({
        success: false,
        message: "No active session found"
      });
    }

    const { client } = clientData;
    const allParticipants: any[] = [];
    let invitedCount = 0;

    // Get target group participants first to check membership
    const targetGroupName = targetGroup.replace('https://t.me/', '');
    const targetParticipants = await client.getParticipants(targetGroupName);
    const targetMemberIds = new Set(targetParticipants?.map(p => p.id) || []);

    // Get participants from each source group
    for (const groupLink of sourceGroups) {
      try {
        const participants = await client.getParticipants(groupLink.replace('https://t.me/', ''));
        
        if (participants) {
          // For each participant, process and invite if needed
          for (const participant of participants) {
            try {
              // Skip if already in target group
              if (targetMemberIds.has(participant.id)) {
                console.log(`${participant.firstName || "User"} is already in target group`);
                continue;
              }

              // Check if user is already in contacts
              const contact = await client.invoke(new Api.contacts.GetContacts({}));
              const contactsResult = contact as any;  // Type assertion for now
              
              const isContact = contactsResult.users?.some((u: any) => u.id === participant.id) || false;

              if (!isContact && participant.phone) {
                // Add to contacts if not already added
                await client.invoke(new Api.contacts.AddContact({
                  id: participant.id,
                  firstName: participant.firstName || "User",
                  lastName: participant.lastName || "",
                  phone: participant.phone,
                  addPhonePrivacyException: true
                }));
                console.log(`Added ${participant.firstName || "User"} to contacts`);
              }

              // Invite to target group
              try {
                await client.invoke(new Api.channels.InviteToChannel({
                  channel: targetGroupName,
                  users: [participant.id]
                }));
                invitedCount++;
                console.log(`Invited ${participant.firstName || "User"} to target group`);
              } catch (inviteError) {
                console.error(`Failed to invite ${participant.firstName || "User"}:`, inviteError);
              }

              allParticipants.push(participant);
            } catch (error) {
              console.error(`Error processing participant ${participant.id}:`, error);
            }
          }
        }
      } catch (error) {
        console.error(`Error getting participants from ${groupLink}:`, error);
      }
    }

    return res.status(200).json({
      success: true,
      message: `Processed ${allParticipants.length} participants, invited ${invitedCount} new members`,
      participants: allParticipants
    });

  } catch (error) {
    console.error('Error getting participants:', error);
    return res.status(500).json({
      success: false,
      message: error instanceof Error ? error.message : 'Failed to get participants'
    });
  }
}

File: tg-bulk-invite/src/pages/api/telegram/getParticipants.py
from flask import request, jsonify
from telethon.tl.functions.channels import InviteToChannelRequest
from connect import app, active_clients

@app.route('/getParticipants', methods=['POST'])
async def get_participants():
    data = request.json
    source_groups = data.get('sourceGroups')
    target_group = data.get('targetGroup')
    session_id = data.get('sessionId')

    try:
        if session_id not in active_clients:
            return jsonify({
                'success': False,
                'message': 'No active session found'
            }), 400

        client = active_clients[session_id]['client']
        all_participants = []
        invited_count = 0

        # Get target group participants
        target_group_name = target_group.replace('https://t.me/', '')
        target_entity = await client.get_entity(target_group_name)
        target_participants = await client.get_participants(target_entity)
        target_member_ids = {p.id for p in target_participants}

        # Process source groups
        for group_link in source_groups:
            try:
                group_name = group_link.replace('https://t.me/', '')
                entity = await client.get_entity(group_name)
                participants = await client.get_participants(entity)

                for participant in participants:
                    if participant.id in target_member_ids:
                        print(f"{participant.first_name or 'User'} is already in target group")
                        continue

                    try:
                        await client(InviteToChannelRequest(
                            channel=target_entity,
                            users=[participant]
                        ))
                        invited_count += 1
                        print(f"Invited {participant.first_name or 'User'} to target group")
                    except Exception as e:
                        print(f"Failed to invite {participant.first_name or 'User'}: {str(e)}")

                    all_participants.append(participant)

            except Exception as e:
                print(f"Error getting participants from {group_link}: {str(e)}")

        return jsonify({
            'success': True,
            'message': f'Processed {len(all_participants)} participants, invited {invited_count} new members',
            'participants': [{'id': p.id, 'firstName': p.first_name} for p in all_participants]
        })

    except Exception as e:
        print(f"Error getting participants: {str(e)}")
        return jsonify({
            'success': False,
            'message': str(e)
        }), 500

if __name__ == '__main__':
    app.run(debug=True) 